### PR TITLE
Cv integration

### PR DESCRIPTION
**Pull request by spencerteetaert**
Fixed projection implementation. 
Added checks for null data. 
Removed unused files.

### CODE REVIEW

File: cv/cv/src/cv_utils/camera_projection.py [hunk 1]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
             self.depth_values += [self.depth_map[key]]
         self.depth_values = np.array(self.depth_values)
 
+        camera_info = rospy.wait_for_message('/zed/zed_node/rgb/camera_info', CameraInfo)
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:29]: The zed camera topic should be `/zed/zed_node/left/camera_info` by default I believe (since it's a stereo camera). Would be good to subscribe too that.

I also wonder if we could combine the left and right frame for better consistency (theoretically everything in the left viewport should match the right viewport except for points not visible). @Jason-Y000 for vis

File: cv/cv/src/cv_utils/camera_projection.py [hunk 2]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
 
-        '''
-        Raw parameters (found in /usr/local/zed/settings)
-        fx = 1400.13
-        fy = 1400.13
-        cx = 1188.06
-        cy = 645.894
+        # Update camera intrinsics to account for the resized new images
+        scale_x = 330.0/camera_info.width
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:32]: Nit: put named constants for 330.0 and 180.0

File: cv/cv/src/cv_utils/camera_projection.py [hunk 3]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
+        scale_y = 180.0/camera_info.height
 
-        image size: 360, 640, 3
+        K = list(camera_info.K)
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:35]: Technically better to not modify the input, but we'll take it if it works

File: cv/cv/src/cv_utils/camera_projection.py (hunk 2 above)
Comment (kajananchinniah) [reply cv/cv/src/cv_utils/camera_projection.py:32]: Would it also be better if we casted `camera_info.width` to a float (since we are treating it as a float here).

The same deal with the height

File: cv/cv/src/cv_utils/camera_projection.py [hunk 4]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
 
-        '''
-        Raw parameters (found in /usr/local/zed/settings)
-        fx = 1400.13
-        fy = 1400.13
-        cx = 1188.06
-        cy = 645.894
+        # Update camera intrinsics to account for the resized new images
+        scale_x = 330.0/camera_info.width
+        scale_y = 180.0/camera_info.height
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:33]: Could you update the height and the width of the camera info too (especially if we are using downstream). Would be good if we add things in the future and for whatever reason rely on the camera info's width/height

File: cv/cv/src/cv_utils/camera_projection.py [hunk 5]
```diff
@@ -3,8 +3,11 @@
...
 import cv2
 import numpy as np
+import rospy
+from image_geometry import PinholeCameraModel
 import rospkg
 
+from sensor_msgs.msg import CameraInfo
 
 class CameraProjection:
     def __init__(self):
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:13]: A slight nit, technically we should be subscribing to camera info continuously and updating the camera model continuously but I'm ok with how it is here since the intrinsics don't change during our execution anyways

File: cv/cv/src/cv_utils/camera_projection.py [hunk 6]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
-        for i in range(len(x_array)):
-            z_map_indeces += [self.keys[str((x_array[i], y_array[i]))]]
-        z_map_indeces = np.array(z_map_indeces)
-
-        x_cam_coords = (x_array - self.ox) / self.fx * self.depth_values[z_map_indeces]
-        y_cam_coords = (y_array - self.oy) / self.fy * self.depth_values[z_map_indeces]
-        z_cam_coords = self.depth_values[z_map_indeces]
+        points = []
+        for i in range(len(pts)):
+            if pts[i][0] < 0 or pts[i][1] < 0:# or pts[i][0] >  or pts[i][1] < 0:
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:75]: Remove dead code

File: cv/cv/src/cv_utils/camera_projection.py (hunk 1 above)
Comment (Jason-Y000) [reply cv/cv/src/cv_utils/camera_projection.py:29]: Should be possible. Think it would be implementing some version of homography to associate between points in the right and left frames

File: cv/cv/src/cv_utils/camera_projection.py [hunk 7]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
+                continue
+            vec = self.camera.projectPixelTo3dRay((pts[i][0], pts[i][1]))
+            z_val = self.depth_map[str((pts[i][0], pts[i][1]))]
+            point3D = [vec[i] * z_val/vec[2] for i in range(3)]
+            points += [point3D]
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:80]: Prefer `.append` (tho I'm not too familar with python2 myself so I'm ok with this if it works)

File: cv/cv/src/cv_utils/camera_projection.py [hunk 8]
```diff
@@ -23,51 +26,60 @@ def __init__(self):
...
-
-        x_cam_coords = (x_array - self.ox) / self.fx * self.depth_values[z_map_indeces]
-        y_cam_coords = (y_array - self.oy) / self.fy * self.depth_values[z_map_indeces]
-        z_cam_coords = self.depth_values[z_map_indeces]
+        points = []
+        for i in range(len(pts)):
+            if pts[i][0] < 0 or pts[i][1] < 0:# or pts[i][0] >  or pts[i][1] < 0:
+                continue
+            vec = self.camera.projectPixelTo3dRay((pts[i][0], pts[i][1]))
+            z_val = self.depth_map[str((pts[i][0], pts[i][1]))]
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/camera_projection.py:78]: I know you mentioned that the str was necessary but is it really?

File: cv/cv/src/cv_utils/zed_wrapper_server.py [hunk 9]
```diff
@@ -13,18 +13,19 @@
...
 
     def run(self):
         rospy.init_node('zed_server_wrapper')
         rospy.Subscriber("image", Image, self.forward_image)
         rospy.spin()
 
     def forward_image(self, data):
-        img = self.bridge.imgmsg_to_cv2(data, desired_encoding='passthrough')
-        self._toRedis(img, "zed/preprocessed")
+        if data != []:
```
Comment (kajananchinniah) [inline cv/cv/src/cv_utils/zed_wrapper_server.py:24]: `is None` instead of `!= []` ?

Comment (kajananchinniah) [reply cv/cv/src/cv_utils/zed_wrapper_server.py:24]: Tho I'm ok with this too

File: cv/lane_detection/src/lane_detection_model_inference.py [hunk 10]
```diff
@@ -28,35 +28,53 @@ def __init__(self):
...
+
+        retval, buffer = cv2.imencode('.png', img)
+        img_bytes = np.array(buffer).tostring()
+
+        encoded = shape + img_bytes
+
+        # Store encoded data in Redis
+        self.redis.set(name,encoded)
+
+        return
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_model_inference.py:77]: Unnecessary return

File: cv/lane_detection/src/lane_detection_wrapper_client.py [hunk 11]
```diff
@@ -6,51 +6,72 @@
...
 from cv.msg import FloatArray, FloatList
 from geometry_msgs.msg import Point
 from cv_utils import camera_projection
 
 class CVWrapperClient:
     def __init__(self):
-        self.redis = redis.Redis(host='127.0.0.1', port=6379, db=0)
+        self.redis = redis.Redis(host='127.0.0.1', port=6379, db=3)
         self.pub = rospy.Publisher('cv/lane_detections', FloatArray, queue_size=10)
+        self.pub_raw = rospy.Publisher('cv/model_output', Image, queue_size=10)
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_wrapper_client.py:21]: Can take this out now ? Although we may wish to have it for the competition for our rosbags (just for debugging)

@1999michael for vis

File: cv/lane_detection/src/lane_detection_wrapper_client.py [hunk 12]
```diff
@@ -6,51 +6,72 @@
...
-                    pt_msg.z = pt[2]
-                    pts_msg += [pt_msg]
-                lane_msg.elements = pts_msg
-                lanes_msgs += [lane_msg]
-
-            msg = FloatArray()
-            msg.lists = lanes_msgs
-            self.pub.publish(msg)
+            lanes = self._fromRedisLanes('lane_detection')
+            mask = self._fromRedisImg('cv/model/output')
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_wrapper_client.py:31]: You may need to double check that these values are not None since in some branches you return None / void

Comment (kajananchinniah) [reply cv/lane_detection/src/lane_detection_wrapper_client.py:31]: I see you do it later on (but you do use mask w/o checking)

I also prefer checking it right away rather than deferring it later

File: cv/lane_detection/src/lane_detection_wrapper_client.py [hunk 13]
```diff
@@ -6,51 +6,72 @@
...
+            img = self.bridge.cv2_to_imgmsg(mask*255, encoding='passthrough')
+            self.pub_raw.publish(img)
+
+
+            if lanes is not None:
+                lanes_msgs = []
+                for lane in lanes:
+                    project_points = self.projection(np.array(lane, dtype=np.int))
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_wrapper_client.py:39]: Is this datatype right? Wouldn't we need to use `np.float` ?

File: cv/lane_detection/src/line_fitting.py [hunk 14]
```diff
@@ -28,13 +28,11 @@ def lane_fitting(points):
     x_width = abs(sorted_points_x[-1][1] - sorted_points_x[0][1])
     y_width = abs(sorted_points_y[-1][0] - sorted_points_y[0][0])
 
-    if (x_width < 15) or (y_width < 15):    # Hard-coded parameter, update maybe
+    if ((x_width > 0.95*y_width) and (x_width < 1.05*y_width) and (x_width < 20 or y_width < 20)) or len(points) < 200:    # Hard-coded parameter, update maybe
         return None
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:32]: I personally dislike returning None as a pattern, but I'm not against it

File: cv/lane_detection/src/line_fitting.py [hunk 15]
```diff
@@ -28,13 +28,11 @@ def lane_fitting(points):
     x_width = abs(sorted_points_x[-1][1] - sorted_points_x[0][1])
     y_width = abs(sorted_points_y[-1][0] - sorted_points_y[0][0])
 
-    if (x_width < 15) or (y_width < 15):    # Hard-coded parameter, update maybe
+    if ((x_width > 0.95*y_width) and (x_width < 1.05*y_width) and (x_width < 20 or y_width < 20)) or len(points) < 200:    # Hard-coded parameter, update maybe
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:31]: Not readable if statement but it's fine if you want to keep it

Also delete the hardcoded parameter, and consider making this into a function that with a descriptive name (e.g. `is_in_roi`) that tells us what it does (but feel free to resolve since it's fine)

File: cv/lane_detection/src/line_fitting.py [hunk 16]
```diff
@@ -28,13 +28,11 @@ def lane_fitting(points):
...
         return None
 
-    # print(sorted_points)
-    pts_added = 0
     total_pts = len(points)
-    num_windows = 20
+    num_windows = 30
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:35]: Capitalize constants: `NUM_WINDOW = 30` is better since it's more clear that this should be immutable

File: cv/lane_detection/src/line_fitting.py [hunk 17]
```diff
@@ -51,99 +49,50 @@ def lane_fitting(points):
...
+    # u = np.linspace(u[0]-EXTRAPOLATE_VALUE,u[-1]+EXTRAPOLATE_VALUE,500)
+    out = interpolate.splev(u,tck)  # out is an array in the form of [[x_points], [y_points]]
 
     if type(out) == list:
-        return [out[0].tolist(), out[1].tolist()]
+        return np.array(out).T.tolist()
     else:
-        return out.tolist()
+        raise NotImplementedError()
+        return out.tolist() # We shouldn't be hitting this, right?
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:71]: Exciting

You may also not want to raise and error and instead just return (@spencerteetaert please confirm what happens if we enter this branch). I think this will take down the process, which we don't want

File: cv/lane_detection/src/lane_detection_wrapper_client.py [hunk 18]
```diff
@@ -6,51 +6,72 @@
...
-    def _fromRedis(self, name):
+    def _fromRedisLanes(self, name):
         """Retrieve Numpy array from Redis key 'n'"""
         encoded = self.redis.get(name)
+        if encoded is None:
+            return None
         lanes = json.loads(encoded)
         return lanes
 
+    def _fromRedisImg(self, name):
```
Comment (kajananchinniah) [inline cv/lane_detection/src/lane_detection_wrapper_client.py:66]: You might want to create a general utils file for all these functions that seem to be used in a lot of different places and have redis be a argument you pass in

File: cv/lane_detection/src/line_fitting.py [hunk 19]
```diff
@@ -51,99 +49,50 @@ def lane_fitting(points):
...
 
-    rows = np.where(mask==1)[0].reshape(-1,1)
-    cols = np.where(mask==1)[1].reshape(-1,1)
+    smoothed_pred = cv2.morphologyEx(mask, cv2.MORPH_OPEN, cv2.getStructuringElement(cv2.MORPH_CROSS,(3,3)))
+    
+    rows = np.where(smoothed_pred==1)[0].reshape(-1,1)
+    cols = np.where(smoothed_pred==1)[1].reshape(-1,1)
     coords = np.concatenate((rows,cols),axis=1)     # (y,x) points
-
+
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:82]: Could you try fixing formatting everywhere in general (we don't have a linter tho so I'm not too strict on this). Just get rid of whitespace tho (your text editor should have settings to do this automatically)

File: cv/lane_detection/src/line_fitting.py [hunk 20]
```diff
@@ -51,99 +49,50 @@ def lane_fitting(points):
...
-        # If we haven't already shown or saved the plot, then we need to
-        # draw the figure first...
-        # fig.canvas.draw()
-
-        # # Now we can save it to a numpy array.
-        # data = np.frombuffer(fig.canvas.tostring_rgb(), dtype=np.uint8)
-        # data = data.reshape(fig.canvas.get_width_height()[::-1] + (3,))
 
         return out
     return None
```
Comment (kajananchinniah) [inline cv/lane_detection/src/line_fitting.py:98]: Might be better to return empty list ? (and add the checking wherever relevant)

File: cv/pothole_detection/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+  std_msgs 
+  geometry_msgs
+  )
+
+# Generate cmake config files via catkin_package
+catkin_package(
+  INCLUDE_DIRS
+  CATKIN_DEPENDS
+  )
+
```
Comment (kajananchinniah) [inline cv/pothole_detection/CMakeLists.txt:16]: Fix end of line (look up how to do it in whaever text editor you are using)

File: cv/pothole_detection/CMakeLists.txt [hunk 22]
```diff
@@ -0,0 +1,16 @@
+cmake_minimum_required(VERSION 3.0.2)
+project(pothole_detection)
+
+# # Find catkin macros and libraries
+find_package(catkin REQUIRED COMPONENTS 
+  rospy 
+  std_msgs 
+  geometry_msgs
+  )
+
```
Comment (kajananchinniah) [inline cv/pothole_detection/CMakeLists.txt:10]: Also on a side note, for the `FloatList` msg, we need to generate the msg so it's viewable in C++

Do we want to do this in here or in VL PR ? @1999michael for vis

File: cv/pothole_detection/launch/pothole_detection_server.launch [hunk 23]
```diff
@@ -0,0 +1,19 @@
+<launch>
+    <!-- Runs CV pipeline in python3 communicating with UDP pipeline -->
+    <node
+        name="pothole_detection_model_inference"
+        pkg="pothole_detection"
+        type="pothole_detection_model_inference.py"
+        output="screen"
+    >
+        <remap from="image" to="/zed/zed_node/rgb/image_rect_color"/>
```
Comment (kajananchinniah) [inline cv/pothole_detection/launch/pothole_detection_server.launch:9]: This should be `left` instead of `rgb` since we don't have a `rgb` topic (in sim). Again, don't know what the real camera does but I anticipate it is similar

File: cv/cv/src/cv_utils/camera_projection.py (hunk 1 above)
Comment (kajananchinniah) [reply cv/cv/src/cv_utils/camera_projection.py:29]: Yea, possibly looking into using a sift detector would work

File: cv/pothole_detection/src/pothole_detection_model_inference.py [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+import rospkg
+from runonnx import run_inference
+
+
+class CVModelInferencer:
+    def __init__(self):
+        self.redis = redis.Redis(host='127.0.0.1', port=6379, db=3)
+        
+        rospack = rospkg.RosPack()
+        model_path = f"{rospack.get_path('pothole_detection')}/models/best.onnx"
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/pothole_detection_model_inference.py:20]: We may need to have a different model name for this

Also we don't have the model commit here so just remember that as a note

File: cv/pothole_detection/src/pothole_detection_model_inference.py [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+
+    def run(self) -> None:
+        raw = self._fromRedis("zed/preprocessed")
+
+        if raw is not None:
+            potholes = run_inference(raw, self.ort_session)
+            if potholes is None:
+                return
+
+            potholes = potholes[0]
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/pothole_detection_model_inference.py:33]: Double check this is not empty also why is this necessary?

File: cv/pothole_detection/src/pothole_detection_wrapper_client.py [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+        encoded = self.redis.get(name)
+        lanes = json.loads(encoded)
+        return lanes
+
+if __name__ == '__main__':
+    wrapper = CVWrapperClient()
+    wrapper.run()
+
+
+
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/pothole_detection_wrapper_client.py:59]: Fix end of line

File: cv/pothole_detection/src/pothole_detection_wrapper_client.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env python2
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/pothole_detection_wrapper_client.py:1]: Comments are gonna be the same as the other

File: cv/lane_detection/src/lane_detection_wrapper_client.py (hunk 13 above)
Comment (spencerteetaert) [reply cv/lane_detection/src/lane_detection_wrapper_client.py:39]: we need to set as int to index our depth map. No way to interpolate values at the momeny

File: cv/pothole_detection/src/runonnx.py [hunk 28]
```diff
@@ -0,0 +1,183 @@
...
+
+    returns an nx6 with (x1, y1, x2, y2, confidence, class)
+    """
+
+    candidates = prediction[..., 4] > conf_thres
+
+    max_det = 300
+    time_limit = 5.0
+
+    t = time.time()
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/runonnx.py:30]: Remove benchmarking

File: cv/lane_detection/src/line_fitting.py (hunk 15 above)
Comment (spencerteetaert) [reply cv/lane_detection/src/line_fitting.py:31]: this is actively being developed, will be pushed again by Jason

File: cv/pothole_detection/src/runonnx.py [hunk 29]
```diff
@@ -0,0 +1,183 @@
...
+    # List of bboxs in format [x1, y1, x2, y2, confs, class]
+    output = []
+
+    for i, det in enumerate(pred):
+        det[:, :4] = scale_coords(
+            img1_shape=(448, 448), coords=det[:, :4], img0_shape=(180, 330)
+        )
+        output.append(det)
+
+    return output
```
Comment (kajananchinniah) [inline cv/pothole_detection/src/runonnx.py:183]: Fix end of file

File: cv/pothole_detection/CMakeLists.txt (hunk 21 above)
Comment (spencerteetaert) [reply cv/pothole_detection/CMakeLists.txt:16]: this is from catkin pkg create? not sure what it's wanting here

File: cv/pothole_detection/launch/pothole_detection_server.launch (hunk 23 above)
Comment (spencerteetaert) [reply cv/pothole_detection/launch/pothole_detection_server.launch:9]: rgb in real zed wrapper corresponds to depth map. This should be like this, the sim is missing the topic. I think it maps to left camera

File: cv/pothole_detection/src/pothole_detection_model_inference.py (hunk 24 above)
Comment (spencerteetaert) [reply cv/pothole_detection/src/pothole_detection_model_inference.py:20]: bad practice to upload large files to git. We have them stored on google drive

**Comment by kajananchinniah** [comment]
I didn't review the pothole code much but it's probably fine. Remind me to review it in the next pass

**Review by kajananchinniah — APPROVED** [review]
Lgtm but we should do another pass most likely with the issues we discussed offline

Also remind me to formally review the pothole code later, but we need this merged for tmrw

File: cv/pothole_detection/src/runonnx.py (hunk 28 above)
Comment (spencerteetaert) [reply cv/pothole_detection/src/runonnx.py:30]: Going to leave but reduce interval. It's a good check to have to prevent this from reducing FPS for if we get lots of detections
